src/Analyser/MutatingScope.php

```diff
@@ -1534,7 +1542,9 @@ private function resolveType(Expr $node): Type
 				}
 
 				$truthyScope = $matchScope->filterByTruthyValue($filteringExpr);
-				$types[] = $truthyScope->getType($arm->body);
+				if (count($arm->conds) !== $impossibleConds) {
+					$types[] = $truthyScope->getType($arm->body);
+				}
```

Comment:
An alternative would be:

```php
$filteringExprType = $matchScope->getType($filteringExpr);

if (!(new ConstantBooleanType(false))->isSuperTypeOf($filteringExprType)->yes()) {
	$truthyScope = $matchScope->filterByTruthyValue($filteringExpr);
	$types[] = $truthyScope->getType($arm->body);
}
```
... but I'm starting to wonder if `filterByTruthyValue` should be handling constant-false expressions better.

Reply:
Ah, this is really unrelated to `filterByTruthyValue`. I do think my suggestion is a bit simpler though :)